File: Trie_of_blooming_directed_graphs_with_agents_and_filters/ruby/root.rb
# Root anchor: Trie_of_blooming_directed_graphs_with_agents_and_filters
# Single entry point for blooming-directed-graph + filter_agents + trigger_process_agent per key.

TrieNodeValue = Struct.new(:graph_ref, :filter_agents_ref, :trigger_process_agent_ref, keyword_init: true)

class TrieOfBloomingDirectedGraphsWithAgentsAndFilters
  def initialize
    @children = {}
    @is_end = false
    @value = nil
  end

  def insert(key, value)
    if key.empty?
      @is_end = true
      @value = value
      return
    end
    c = key[0]
    @children[c] ||= TrieOfBloomingDirectedGraphsWithAgentsAndFilters.new
    @children[c].insert(key[1..], value)
  end

  def get(key)
    if key.empty?
      return @is_end ? @value : nil
    end
    c = key[0]
    child = @children[c]
    child&.get(key[1..])
  end

  def starts_with?(prefix)
    return true if prefix.empty?
    c = prefix[0]
    child = @children[c]
    child&.starts_with?(prefix[1..]) || false
  end
end

# Global root anchor object (Trie_of_blooming_directed_graphs_with_agents_and_filters).
Trie_of_blooming_directed_graphs_with_agents_and_filters = TrieOfBloomingDirectedGraphsWithAgentsAndFilters.new

File: Trie_of_blooming_directed_graphs_with_agents_and_filters/csharp/Root.cs
// Root anchor: Trie_of_blooming_directed_graphs_with_agents_and_filters
// Single entry point for blooming-directed-graph + filter_agents + trigger_process_agent per key.

using System.Collections.Generic;

namespace TrieRootAnchor
{
    public class TrieNodeValue
    {
        public object? GraphRef { get; set; }
        public object? FilterAgentsRef { get; set; }
        public object? TriggerProcessAgentRef { get; set; }
    }

    public class TrieOfBloomingDirectedGraphsWithAgentsAndFilters
    {
        private readonly Dictionary<char, TrieOfBloomingDirectedGraphsWithAgentsAndFilters> _children = new();
        private bool _isEnd;
        private TrieNodeValue? _value;

        public void Insert(string key, TrieNodeValue value)
        {
            if (string.IsNullOrEmpty(key))
            {
                _isEnd = true;
                _value = value;
                return;
            }
            var c = key[0];
            if (!_children.TryGetValue(c, out var child))
            {
                child = new TrieOfBloomingDirectedGraphsWithAgentsAndFilters();
                _children[c] = child;
            }
            child.Insert(key.Substring(1), value);
        }

        public TrieNodeValue? Get(string key)
        {
            if (string.IsNullOrEmpty(key))
                return _isEnd ? _value : null;
            if (!_children.TryGetValue(key[0], out var child))
                return null;
            return child.Get(key.Substring(1));
        }

        public bool StartsWith(string prefix)
        {
            if (string.IsNullOrEmpty(prefix)) return true;
            if (!_children.TryGetValue(prefix[0], out var child))
                return false;
            return child.StartsWith(prefix.Substring(1));
        }
    }

    /// <summary>Global root anchor (Trie_of_blooming_directed_graphs_with_agents_and_filters).</summary>
    public static class Root
    {
        public static readonly TrieOfBloomingDirectedGraphsWithAgentsAndFilters Trie_of_blooming_directed_graphs_with_agents_and_filters =
            new TrieOfBloomingDirectedGraphsWithAgentsAndFilters();
    }
}
